docs/testplan/ECN_test.md

```diff
@@ -0,0 +1,114 @@
+# SONiC Switch ECN Test
+
+- [SONiC Switch ECN Test](#sonic-switch-ecn-test)
+  - [Overview](#overview)
+  - [Test Objective](#test-objective)
+  - [Test Setup](#test-setup)
+  - [Test Cases](#test-cases)
+    - [ECN marking at egress](#ecn-marking-at-egress)
+    - [ECN marking accuracy](#ecn-marking-accuracy)
+
+## Overview
+
+In a traditional TCP/IP network, congestion is detected through packet loss. When network congestion occurs, routers and switches drop packets when their buffers fill up.
+Explicit Congestion Notification (ECN) is an enhancement to TCP/IP congestion control that allows congestion to be signaled without dropping packets. Instead of relying on packet loss, ECN uses marking within the IP header to indicate congestion. The process works as follows:
+
+- ECN-Capable Transport (ECT) Marking: The sender marks packets as ECN-capable (ECT) in the IP header.
+- Congestion Notification: If a router or switch experiences congestion, instead of dropping packets, it marks the Congestion Experienced (CE) bit in the IP header.
+- Receiver Acknowledgment: The receiver detects the CE mark and informs the sender via an ECN-Echo (ECE) flag in the TCP acknowledgment.
+- Sender Response: Upon receiving an ECN-Echo, the sender reduces its congestion window (CWND) to slow down transmission, similar to traditional congestion control, but without losing packets.
+
+This proactive mechanism improves network efficiency by preventing packet drops and reducing retransmission overhead. The primary algorithm governing ECN behavior is the Random Early Detection (RED) with ECN marking. RED dynamically marks packets with ECN-CE based on the queue depth. As the queue length increases, the marking probability rises, signaling congestion before significant queuing delays occur. The RED algorithm is defined by three key parameters:
+
+- 𝐾min – the minimum threshold
+- 𝐾max – the maximum threshold
+- 𝑃max – the maximum marking (or dropping) probability
+
+The marking probability is determined as follows:
+
+- When the instantaneous queue length is below Kmin, the marking probability is 0%.
+- When the queue length exceeds Kmax, the marking probability reaches 100%.
+- Otherwise, the marking probability follows the formula: Marking Probability = (queue_length − 𝐾min)/(𝐾max − 𝐾min) × 𝑃max
+
+The figure in ECN-test-plan.md illustrates the theoretical ECN marking probability as a function of queue length. Commodity switches can implement RED at ingress (when packets are enqueued into the switch buffer) or at egress (when packets are dequeued from the buffer). Compared to ingress RED/ECN, egress RED/ECN provides lower feedback delay, leading to faster congestion response.
+
+## Test Objective
+
+The test aims to validate the ECN behaviors of the SONiC DUT based on the Random Early Detection (RED) algorithm.
+
+## Test Setup
+
+The test is designed to be topology-agnostic, meaning it does not assume or impose a specific network connection.
```

Comment:
Hi Mei, this test actually might be much easier to do using a single DUT and also will be more reliable.

First thing is that we don't have to replace the rpc syncd anymore.

Second thing is that the multiple layers of switch might all introduced marked packets, which results in unexpected result.

Since we are still stopping the queue, we might not be required to do anything stressing the switch, so a single DUT might look already enough in this case.